src/iocp/mod.rs

```diff
@@ -0,0 +1,821 @@
+    pub(super) fn supports_level(&self) -> bool {
+        true
+    }
+
+    /// Whether this poller supports edge-triggered events.
+    pub(super) fn supports_edge(&self) -> bool {
+        false
+    }
+
+    /// Add a new source to the poller.
+    pub(super) fn add(&self, socket: RawSocket, interest: Event, mode: PollMode) -> io::Result<()> {
+        log::trace!(
+            "add: handle={:?}, sock={}, ev={:?}",
+            self.port,
+            socket,
+            interest
+        );
+
+        // We don't support edge-triggered events.
+        if matches!(mode, PollMode::Edge) {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "edge-triggered events are not supported",
+            ));
+        }
+
+        // Create a new packet.
+        let socket_state = {
+            let state = SocketState {
+                socket,
+                base_socket: base_socket(socket)?,
+                interest,
+                interest_error: true,
+                afd: self.afd_handle()?,
+                mode,
+                waiting_on_delete: false,
+                status: SocketStatus::Idle,
+            };
+
+            Arc::pin(IoStatusBlock::from(PacketInner::Socket {
+                packet: UnsafeCell::new(AfdPollInfo::default()),
+                socket: Mutex::new(state),
+            }))
+        };
+
+        // Keep track of the source in the poller.
+        {
+            let mut sources = self.sources.write().unwrap_or_else(|e| e.into_inner());
+
+            match sources.entry(socket) {
+                Entry::Vacant(v) => {
+                    v.insert(Pin::<Arc<_>>::clone(&socket_state));
+                }
+
+                Entry::Occupied(_) => {
+                    return Err(io::Error::from(io::ErrorKind::AlreadyExists));
+                }
+            }
+        }
+
+        // Update the packet.
+        self.update_packet(socket_state)
+    }
+
+    /// Update a source in the poller.
+    pub(super) fn modify(
+        &self,
+        socket: RawSocket,
+        interest: Event,
+        mode: PollMode,
+    ) -> io::Result<()> {
+        log::trace!(
+            "modify: handle={:?}, sock={}, ev={:?}",
+            self.port,
+            socket,
+            interest
+        );
+
+        // We don't support edge-triggered events.
+        if matches!(mode, PollMode::Edge) {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "edge-triggered events are not supported",
+            ));
+        }
+
+        // Get a reference to the source.
+        let source = {
+            let sources = self.sources.read().unwrap_or_else(|e| e.into_inner());
+
+            match sources.get(&socket) {
+                Some(s) => s.clone(),
+                None => {
+                    return Err(io::Error::from(io::ErrorKind::NotFound));
+                }
+            }
```

Comment:
this might be cleaner done via chaining, e.g.
```suggestion
            sources.get(&socket)
                .as_ref()
                .cloned()
                .ok_or_else(|| io::Error::from(io::ErrorKind::NotFound))?
```